### DISCUSSION TITLE
Hang when using custom function on parameters

### DISCUSSION THREAD

**Discussion by psmaragdis** [apache/incubator-mxnet#15844] [Q&A]
## Description
Custom functions made using autograd.Function hang if the functions's input is used in self.saved_tensors and that function is applied on a layer parameter. It seems to work fine when this function is applied on non-parameters in a layer.

## Environment info (Required)

```
----------Python Info----------
Version      : 3.6.6
Compiler     : GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)
Build        : ('default', 'Jun 28 2018 11:07:29')
Arch         : ('64bit', '')
------------Pip Info-----------
Version      : 19.0.3
Directory    : /usr/local/anaconda/lib/python3.6/site-packages/pip
----------MXNet Info-----------
Version      : 1.5.0
Directory    : /usr/local/anaconda/lib/python3.6/site-packages/mxnet
Commit Hash   : 75a9e187d00a8b7ebc71412a02ed0e3ae489d91f
Library      : ['/usr/local/anaconda/lib/python3.6/site-packages/mxnet/libmxnet.so']
Build features:
✖ CUDA
✖ CUDNN
✖ NCCL
✖ CUDA_RTC
✖ TENSORRT
✔ CPU_SSE
✔ CPU_SSE2
✔ CPU_SSE3
✔ CPU_SSE4_1
✔ CPU_SSE4_2
✖ CPU_SSE4A
✔ CPU_AVX
✖ CPU_AVX2
✖ OPENMP
✖ SSE
✖ F16C
✖ JEMALLOC
✖ BLAS_OPEN
✖ BLAS_ATLAS
✖ BLAS_MKL
✖ BLAS_APPLE
✔ LAPACK
✖ MKLDNN
✔ OPENCV
✖ CAFFE
✖ PROFILER
✔ DIST_KVSTORE
✖ CXX14
✖ INT64_TENSOR_SIZE
✔ SIGNAL_HANDLER
✖ DEBUG
----------System Info----------
Platform     : Darwin-18.7.0-x86_64-i386-64bit
system       : Darwin
node         : kados
release      : 18.7.0
version      : Darwin Kernel Version 18.7.0: Thu Jun 20 18:42:21 PDT 2019; root:xnu-4903.270.47~4/RELEASE_X86_64
----------Hardware Info----------
machine      : x86_64
processor    : i386
b'machdep.cpu.brand_string: Intel(R) Xeon(R) CPU E5-1650 v2 @ 3.50GHz'
b'machdep.cpu.features: FPU VME DE PSE TSC MSR PAE MCE CX8 APIC SEP MTRR PGE MCA CMOV PAT PSE36 CLFSH DS ACPI MMX FXSR SSE SSE2 SS HTT TM PBE SSE3 PCLMULQDQ DTES64 MON DSCPL VMX SMX EST TM2 SSSE3 CX16 TPR PDCM SSE4.1 SSE4.2 x2APIC POPCNT AES PCID XSAVE OSXSAVE TSCTMR AVX1.0 RDRAND F16C'
b'machdep.cpu.leaf7_features: RDWRFSGS SMEP ERMS MDCLEAR IBRS STIBP L1DF SSBD'
b'machdep.cpu.extfeatures: SYSCALL XD 1GBPAGE EM64T LAHF RDTSCP TSCI'
----------Network Test----------
Setting timeout: 10
Timing for MXNet: https://github.com/apache/incubator-mxnet, DNS: 0.0572 sec, LOAD: 0.7426 sec.
Timing for Gluon Tutorial(en): http://gluon.mxnet.io, DNS: 0.0788 sec, LOAD: 0.3005 sec.
Timing for Gluon Tutorial(cn): https://zh.gluon.ai, DNS: 0.1770 sec, LOAD: 0.2624 sec.
Timing for FashionMNIST: https://apache-mxnet.s3-accelerate.dualstack.amazonaws.com/gluon/dataset/fashion-mnist/train-labels-idx1-ubyte.gz, DNS: 0.0256 sec, LOAD: 0.2692 sec.
Timing for PYPI: https://pypi.python.org/pypi/pip, DNS: 0.0232 sec, LOAD: 0.3674 sec.
Timing for Conda: https://repo.continuum.io/pkgs/free/, DNS: 0.0442 sec, LOAD: 0.0687 sec.

```

Package used (Python/R/Scala/Julia):
I'm using Python

## Error Message:
No error, program hangs

## Minimum reproducible example
```
import mxnet as mx
import mxnet.ndarray as nd
import mxnet.gluon.nn as nn

# Custom function with saved tensor being the input (function doesn't matter)
class myfun( mx.autograd.Function):
    def forward( self, x):
        self.save_for_backward( x)
        return x*x

    def backward( self, dx):
        x, = self.saved_tensors
        return dx * 2*x


# A layer using that function on its parameter
class mylinear( nn.Block):
    def __init__( self, ins, outs):
        super( mylinear, self).__init__()
        self.weight = self.params.get( 'weight', shape=(outs,ins))

    def forward( self, x):
        # This hangs
        w = myfun()( self.weight.data())

        # But this works
        # w = self.weight.data()*self.weight.data()

        return nd.dot( x, w.transpose())


# Training loop
def trainit( model, bt=1024):
    model.initialize()
    opt = mx.gluon.Trainer( model.collect_params(), 'SGD', {'learning_rate':1e-3})
    for ep in range( 10):
        x = nd.random.normal( shape=(bt,10))
        y = nd.random.normal( shape=(bt,2))

        # Get loss and update
        with mx.autograd.record( train_mode=True):
            l = (model( x) - y).abs().mean()
            l.backward()
        opt.step( bt)

        # Stuck here
        print( 'Epoch:', ep, 'loss:', l.asscalar())

trainit( mylinear( 10, 2))
```

## Steps to reproduce
Run the code above, it should hang after the first iteration.

## What have you tried to solve it?
(not solutions, but hopefully helpful observations)

If I change the squaring of the layer parameter to the inline function (commented out in my code) it does not hang.  But this gets in the way of specifying custom backward steps.

If I use a custom function where the output of the function is stored for the backward step it does not hang, but if the derivative cannot be computed as such there is no solution that I see.

If I apply this function on a non-parameter, e.g. an input, it works fine.

What I compute in the function doesn't matter, as long as I use the input in the backwards step

**Comment by vdantu** [comment]
@mxnet-label-bot add [question]

**Comment by arcadiaphy** [comment]
@psmaragdis Looks like we've forgot to add read dependencies on saved tensors, I'm working on a PR to fix it.
